File: backoffice/src/app/core/utils/media-url.ts
import { environment } from '../../../environments/environment';

/**
 * Résout une URL média (image de post, avatar) pour l'affichage.
 *
 * Les fichiers sont hébergés par l'API sous `/uploads/…`. On les ré-ancre
 * toujours sur l'**origine API configurée** — même si la base contient une
 * ancienne URL absolue avec un hôte périmé (ex. une IP LAN `http://192.168.x.x:5199`
 * gravée quand un post a été créé depuis le mobile). Cela « répare » l'affichage
 * sans migration de données. Les autres URLs absolues (images externes : picsum,
 * unsplash, pravatar…) passent telles quelles.
 */
export function resolveMediaUrl(url: string | null | undefined): string {
  if (!url) return '';

  // Origine de l'API (apiUrl est « …/api » ; les uploads sont servis à la racine).
  const origin = environment.apiUrl.replace(/\/api\/?$/, '').replace(/\/+$/, '');

  // Fichier hébergé par l'API : on ne garde que le chemin à partir de /uploads/
  // et on le rattache à l'origine courante (répare les hôtes périmés).
  const uploadsIdx = url.indexOf('/uploads/');
  if (uploadsIdx >= 0) return `${origin}${url.slice(uploadsIdx)}`;

  // URL déjà absolue et externe → inchangée.
  if (/^https?:\/\//i.test(url)) return url;

  // Chemin relatif quelconque → préfixé par l'origine API.
  return url.startsWith('/') ? `${origin}${url}` : `${origin}/${url}`;
}

File: mobile/lib/core/utils/media_url.dart
import '../config/app_config.dart';

/// Résout une URL média renvoyée par l'API.
///
/// Les fichiers uploadés sont stockés en **chemin relatif** (`/uploads/…`, sans
/// nom de domaine) afin que la base reste portable si l'hôte change (dev → prod).
/// On les préfixe donc avec la base API configurée du client. Les URLs déjà
/// absolues (http/https — images externes seed, avatars pravatar, etc.) passent
/// telles quelles. Renvoie `''` si l'entrée est nulle/vide.
String resolveMediaUrl(String? url) {
  if (url == null || url.isEmpty) return '';
  final base = AppConfig.apiBaseUrl.replaceAll(RegExp(r'/+$'), '');

  // Fichier hébergé par l'API : on ré-ancre le chemin à partir de « /uploads/ »
  // sur la base API courante — répare aussi les anciennes URLs absolues dont
  // l'hôte est périmé (ex. IP LAN `http://192.168.x.x:5199` gravée en base).
  final uploadsIdx = url.indexOf('/uploads/');
  if (uploadsIdx >= 0) return '$base${url.substring(uploadsIdx)}';

  // URL déjà absolue et externe (picsum, unsplash, pravatar…) → inchangée.
  if (url.startsWith('http://') || url.startsWith('https://')) return url;

  // Chemin relatif quelconque → préfixé par la base API.
  return url.startsWith('/') ? '$base$url' : '$base/$url';
}
